### ISSUE TITLE
NUnit 2021 has a false positive when mixing generics and `System.Enum`

### ISSUE THREAD

**Issue by RenderMichael** [nunit/nunit.analyzers#517]
The following minimal repro gives a false positive
``` csharp
public static void VerificationTest<T>(T? actual, T? expected)
{
    if (actual is System.Enum e)
    {
        Assert.That(e, Is.EqualTo(expected)); // NUnit2021: The EqualTo constraint always fails as the actual and the expected value cannot be equal.
    }
}
```

**Comment by manfred-brands** [comment]
The analyzer only sees the types `System.Enum` and `T?`. It is outside the scope of the analyzer to find out that you actually assigned `e` from a `T?`. In this case it might be right above it, but in other cases it might be way more complex.
Note that even if `actual` is an `Enum`, when `T` is `object` then `expected` could still be anything.

```csharp
object actual = true;
object expected = 42;
VerificationTest(actual, expected)
```

**Comment by RenderMichael** [comment]
@manfred-brands I see what you mean, we can't assume `typeof(T).IsEnum` is true. That check should be added in my code.

But `NUnit2021`'s claim that `EqualTo constraint always fails` is not correct here, is there any way to fix that?

**Comment by manfred-brands** [comment]
We could change the message from `always` to `likely` but I think your case is more an exception and even will go away if you verify that `expected` has the same type as `actual`. 
If you remove the test and simply call the `Assert` the error goes away as well as both arguments are declared with the same `T?`.

**Comment by RenderMichael** [comment]
I tried two things, neither of which fixed the analyzer. 
```csharp
public static void VerificationTest<T>(T? actual, T? expected)
{
    if (typeof(T).IsEnum && actual is System.Enum e)
    {
        Assert.That(e, NUnit.Framework.Is.EqualTo(expected)); // NUnit2021: The EqualTo constraint always fails as the actual and the expected value cannot be equal.
    }
}
```

```csharp
public static void VerificationTest<T>(T? actual, T? expected)
{
    if (actual is System.Enum e)
    {
        Assert.That(e, NUnit.Framework.Is.TypeOf<T>().And.EqualTo(expected)); // NUnit2021: The EqualTo constraint always fails as the actual and the expected value cannot be equal.
    }
}
```

Maybe at the very least `Is.TypeOf<T>()` should satisfy the analyzer?

**Comment by manfred-brands** [comment]
Unfortunately not. The Analyzer only see `e` as being of type `System.Enum` and `expected` of type `T`.

You would have to go to:
```csharp
public static void VerificationTest<T>(T? actual, T? expected)
{
    if (actual is System.Enum eActual &&
        expected is System.Enum eExpected)
    {
        Assert.That(eActual, Is.EqualTo(eExpected));
    }
}
```

But even here, you hide the analyzer warning, but _actual_ and _expected_ could still be of different enumeration:

```csharp
        private enum NoYes {  No, Yes };
        private enum FalseTrue {  False, True };

        [Test]
        public void ActualTest()
        {
            VerificationTest<System.Enum>(NoYes.Yes, FalseTrue.True);
        }
```

I still not sure why you single out a specific type instead of just calling `Is.EqualTo` on the values directly.

**Comment by RenderMichael** [comment]
This example is a minimal repro, in my real code I have special handling of enums for logging purposes. 

I found a solution what works for me:
```csharp
public static void VerificationTest<T>(T? actual, T? expected)
{
    if (actual is System.Enum)
    {
        Assert.That(actual, Is.EqualTo(expected));
    }
}
```

I wanted to make the generic type of the call `Assert.That<System.Enum>()` but from what I see, it doesn't seem to affect the results? The above `Assert.That<T>()` doesn't change the outcome, the logging, or anything else as far as I can tell.
